File: 01.DFS递归/leetcode_1306_canReach.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
private:
    int n;
    vector<int> visited, Arr;
    
    bool dfs(int indx){
        if (indx < 0 || indx >= n || visited[indx]){
            return false;
        }
        if (Arr[indx] == 0){
            return true;
        }
        visited[indx] = 1;
        return dfs(indx+Arr[indx]) | dfs(indx-Arr[indx]);
    }

public:
    bool canReach(vector<int>& arr, int start) {
        Arr = arr;
        n = arr.size();
        visited.resize(n, 0);
        return dfs(start);
    }
};

int main(){
    Solution s;
    vector<int> arr = {4,2,3,0,3,1,2};
    int start = 5;
    cout << s.canReach(arr, start) << endl;
}

File: 01.DFS递归/leetcode_1306_canReach.py
from typing import List

class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        n = len(arr)
        visited = [0]*n
        def dfs(indx):
            if indx >= n or indx < 0 or visited[indx]:
                return False
            if arr[indx] == 0:
                return True
            visited[indx] = 1
            return dfs(indx+arr[indx]) | dfs(indx-arr[indx])
        return dfs(start)

if __name__ == "__main__":
    s = Solution()
    arr = [4,2,3,0,3,1,2]
    start = 5
    print(s.canReach(arr, start))
